### PR TITLE
solution

### PR DESCRIPTION
**Pull request by abcdesign1**
[DEMO LINK](https://vadym4che.github.io/react_dynamic-list-of-todos/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -8,7 +7,70 @@ import { TodoFilter } from './components/TodoFilter';
...
+        if (todosFromServer) {
+          setTodos([...todosFromServer]);
+          setLoadingTodoList(false);
+        }
+      });
+  }, []);
+
+  const filteredByStatus = filterTodoByStatus(filterType, todos);
+
+  const todosToShow = filterTodosBySearchQuery(searchQuery, filteredByStatus);
```
Comment (illiarerun) [inline src/App.tsx:70]: use `useMemo` for this to avoid recalculation on EVERY rerender

File: src/App.tsx [hunk 2]
```diff
@@ -8,7 +7,70 @@ import { TodoFilter } from './components/TodoFilter';
...
+    setSelectedTodoId(id);
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then((todosFromServer) => {
+        if (todosFromServer) {
+          setTodos([...todosFromServer]);
+          setLoadingTodoList(false);
+        }
```
Comment (illiarerun) [inline src/App.tsx:64]: I don't see any error handling

File: src/api.ts [hunk 3]
```diff
@@ -24,4 +24,14 @@ function get<T>(url: string): Promise<T> {
...
 
+// export const getUser = (userId: number) => {
+//   return get<User[]>('/users')
+//     .then((users) => {
+//       // eslint-disable-next-line no-console
+//       console.log(users, userId);
+
+//       return users.find((user) => user.id === userId);
+//     });
+// };
```
Comment (illiarerun) [inline src/api.ts:35]: remove commented code

**Review by illiarerun — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 1 above)
Comment (vadym4che) [reply src/App.tsx:70]: done:
recalculating now happening only if exactly parameter that function and its result depend on is changed.

File: src/App.tsx (hunk 2 above)
Comment (vadym4che) [reply src/App.tsx:64]: done
:
handled in accurate way with bulma notification class :)

File: src/api.ts (hunk 3 above)
Comment (vadym4che) [reply src/api.ts:35]: removed

File: src/App.tsx [hunk 4]
```diff
@@ -8,7 +7,77 @@ import { TodoFilter } from './components/TodoFilter';
...
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
+
+import { FilterType } from './types/FilterType';
+
+const getTodoById = (selectedId: number, todosArr: Todo[]) => {
+  return todosArr.find(({ id }) => id === selectedId);
+};
+
+const filterTodoByStatus = (status: string, todosArr: Todo[]) => {
```
Comment (sergii-nosachenko) [inline src/App.tsx:19]:
```suggestion
const filterTodoByStatus = (filterType: FilterType, todosArr: Todo[]) => {
```

File: src/App.tsx [hunk 5]
```diff
@@ -8,7 +7,77 @@ import { TodoFilter } from './components/TodoFilter';
...
+      filteredTodos = todosArr.filter(({ completed }) => !completed);
+      break;
+    case 'completed':
+      filteredTodos = todosArr.filter(({ completed }) => completed);
+      break;
+    default:
+      filteredTodos = todosArr;
+  }
+
+  return filteredTodos;
```
Comment (sergii-nosachenko) [inline src/App.tsx:37]:
```suggestion
  switch (filterType) {
    case FilterType.Active:
      return todosArr.filter(({ completed }) => !completed);
    case FilterType.Completed:
      return todosArr.filter(({ completed }) => completed);
    default:
      return todosArr;
  }
```

File: src/App.tsx [hunk 6]
```diff
@@ -8,7 +7,77 @@ import { TodoFilter } from './components/TodoFilter';
...
+
+const filterTodosBySearchQuery = (query: string, todosArr: Todo[]) => {
+  return query === ''
+    ? todosArr
+    : todosArr.filter(({ title }) => title.toLocaleLowerCase().includes(query));
+};
+
 export const App: React.FC = () => {
+  const [loadingTodoList, setLoadingTodoList] = useState(true);
+  const [loadingError, setLoadingError] = useState(false);
```
Comment (sergii-nosachenko) [inline src/App.tsx:48]:
```suggestion
  const [isLoadingTodoList, setIsLoadingTodoList] = useState(true);
  const [isLoadingError, setIsLoadingError] = useState(false);
```

File: src/App.tsx [hunk 7]
```diff
@@ -17,18 +86,48 @@ export const App: React.FC = () => {
...
+                  selectedTodoId={selectedTodoId}
+                  onSelect={onSelect}
+                />
+              )}
+
+              {!loadingTodoList && loadingError && (
+                <div className="notification is-danger is-light">
+                  {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
+                  <button
+                    type="button"
```
Comment (sergii-nosachenko) [inline src/App.tsx:112]:
```suggestion
                  <button
                    aria-label="Close notification"
                    type="button"
```

File: src/App.tsx [hunk 8]
```diff
@@ -17,18 +86,48 @@ export const App: React.FC = () => {
...
+                  Try to reload page, or close this massage to auto-reload.
+                </div>
+              )}
             </div>
           </div>
         </div>
       </div>
-
-      <TodoModal />
+      {selectedTodo && (
```
Comment (sergii-nosachenko) [inline src/App.tsx:125]:
```suggestion

      {selectedTodo && (
```

File: src/components/TodoFilter/TodoFilter.tsx [hunk 9]
```diff
@@ -1,34 +1,86 @@
...
+            onChange={handleSelectFilter}
+          >
+            {Object.entries(FilterType)
+              .map((key) => {
+                return (
+                  <option
+                    key={key[0]}
+                    value={key[1]}
+                  >
+                    {key[0]}
```
Comment (sergii-nosachenko) [inline src/components/TodoFilter/TodoFilter.tsx:49]:
```suggestion
              .map(([key, value]) => {
                return (
                  <option
                    key={key}
                    value={value}
                  >
                    {key}
```

File: src/components/TodoFilter/TodoFilter.tsx [hunk 10]
```diff
@@ -1,34 +1,86 @@
...
+        <span className="icon is-left">
+          <i className="fas fa-magnifying-glass" />
+        </span>
+
+        <span className="icon is-right" style={{ pointerEvents: 'all' }}>
+
+          {isResetButtonVisible && (
+            /* eslint-disable-next-line jsx-a11y/control-has-associated-label */
+            <button
+              data-cy="clearSearchButton"
```
Comment (sergii-nosachenko) [inline src/components/TodoFilter/TodoFilter.tsx:75]:
```suggestion
            <button
              aria-label="Clear search"
              data-cy="clearSearchButton"
```

File: src/components/TodoModal/TodoModal.tsx [hunk 11]
```diff
@@ -15,30 +48,37 @@ export const TodoModal: React.FC = () => {
...
               data-cy="modal-header"
             >
-              Todo #2
+              Todo #
+              {id}
             </div>
 
             {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
             <button
               type="button"
```
Comment (sergii-nosachenko) [inline src/components/TodoModal/TodoModal.tsx:57]:
```suggestion
            <button
              aria-label="Close modal"
              type="button"
```

**Review by sergii-nosachenko — APPROVED** [review]

**Pull request closed without merging**